### PR TITLE
planner: optimize `ifnull(not-null-column, ...)` to a cast instead of eliminating it. tidb-test=pr/2098

### PR DESCRIPTION
**Pull request by LittleFall**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/41734

Problem Summary:  see https://github.com/pingcap/tidb/issues/41734#issuecomment-1449475900

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -1883,7 +1883,7 @@ func (er *expressionRewriter) betweenToExpression(v *ast.BetweenExpr) {
 // Otherwise it should return false to indicate (the caller) that original behavior needs to be performed.
 func (er *expressionRewriter) rewriteFuncCall(v *ast.FuncCallExpr) bool {
 	switch v.FnName.L {
-	// when column is not null, ifnull on such column is not necessary.
+	// when column is not null, ifnull on such column can be optimized to a cast.
```
Comment (hawkingrei) [inline planner/core/expression_rewriter.go:1886]: Can you add some tests for it?

Comment (LittleFall) [reply planner/core/expression_rewriter.go:1886]: adding

**Review by hawkingrei — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]
LGTM for the bugfix.
But I think we should be cautious for cherry-picking because it may cause plan changes, and there are also pitfalls in current tidb's handling of type conversions.

**Comment by LittleFall** [comment]
> LGTM for the bugfix. But I think we should be cautious for cherry-picking because it may cause plan changes, and there are also pitfalls in current tidb's handling of type conversions.

highly agree

**Pull request merged**